### PR TITLE
[TS migration] Migrate `SettingsWallet` page to TypeScript

### PR DESCRIPTION
**Pull request by tienifr**
### Details

TS migration `SettingsWallet`.

### Fixed Issues

$ https://github.com/Expensify/App/issues/32003
PROPOSAL:

#### `ExpensifyCardPage`

1. Mock data set in console:

```javascript
Onyx.merge(`cardList`, {
    234523452345: {
        cardID: 234523452345,
        state: 3,
        bank: 'Expensify Card',
        availableSpend: 10000,
        domainName: 'Expensify',
        lastFourPAN: '2345',
        isVirtual: true,
        cardholderFirstName: 'Test',
        cardholderLastName: 'Test',
    },
    234523452346: {
        cardID: 234523452346,
        state: 4,
        bank: '1000',
        availableSpend: 10000,
        domainName: 'Expensify',
        lastFourPAN: '2346',
        isVirtual: false,
    },
});
```

2. Go to `/settings/wallet/card/Expensify`
3. Verify card preview match the above mock and card number is hidden
4. Hard code card details
5. Verify card details match; card number is NOT bold; and expiration date is in the format `MM/YYYY`

#### `ReportVirtualCardFraudPage`

6. Press Report virtual card fraud >> Deactivate card
7. Verify API is called with correct params

#### `ReportCardLostPage`

8. Press Report physical card loss / damage >> Next
9. Verify an error shows require you to choose reason
10. Choose a reason and complete the flow
11. Verify API is called with correct params

#### `AddDebitCardPage`

12. Go to `/settings/wallet/add-debit-card`
13. Fill in invalid values in some fields
14. Verify errors show
15. Leave some fields unfilled >> Next
16. Verify errors show

#### `ActivatePhysicalCardPage`

**Precondition**: Have an unactivated card.

17. Go to `/settings/wallet/card/Expensify/activate`
18. Fill in last four digits of card number
19. Press Activate physical card
20. Verify API is called with correct params
21. Verify error shows `That didn't match the last four digits on the card. Please try again.`

#### `TransferBalancePage`, `ChooseTransferAccountPage`

22. Mock Onyx data

<details>

```javascript
Onyx.merge('bankAccountList', {
    1: {
        title: `Generic bank Checking`,
        description: 'Account ending in 5087',
        methodID: '1',
        key: `bankAccount-1`,
        accountType: 'bankAccount',
        accountData: {
            accountNumber: 'XXXXX5087',
            additionalData: {
                bankName: `Generic bank`,
            },
            state: 'OPEN',
            type: 'PERSONAL',
        },
        isDefault: false,
    },
    2: {
        title: `Generic bank Checking 2`,
        description: 'Account ending in 5123',
        methodID: '2',
        key: `bankAccount-2`,
        accountType: 'bankAccount',
        accountData: {
            accountNumber: 'XXXXX5088',
            additionalData: {
                bankName: `Generic bank`,
            },
            state: 'OPEN',
            type: 'BUSINESS',
        },
        isDefault: false,
    },
});

Onyx.merge('fundList', {
    4: {
        title: 'Test card',
        description: 'Card ending in 4242',
        methodID: 4,
        key: 'card-4',
        accountType: 'debitCard',
        accountData: {
            additionalData: {
                isBillingCard: false,
                isP2PDebitCard: true,
            },
            addressName: 'Test cardList card',
            addressState: 'US',
            addressStreet: 'Street',
            addressZip: 75092,
            cardMonth: 12,
            cardNumber: '424242XXXXXX4242',
            cardYear: 2026,
            created: '2023-01-13 22:27:23',
            currency: 'USD',
            fundID: 4,
        },
        isDefault: false,
    },
    5: {
        title: 'Test card 2',
        description: 'Card ending in 0000',
        methodID: 5,
        key: 'card-5',
        accountType: 'debitCard',
        accountData: {
            additionalData: {
                isBillingCard: false,
                isP2PDebitCard: true,
            },
            addressName: 'Nowhere',
            addressState: 'US',
            addressStreet: 'Street',
            addressZip: 75092,
            cardMonth: 12,
            cardNumber: '424242XXXXXX0000',
            cardYear: 2026,
            created: '2023-01-13 22:27:23',
            currency: 'USD',
            fundID: 5,
        },
        isDefault: false,
    },
});
```

</details>

23. Go to `settings/wallet/transfer-balance`
24. Mock Onyx

```javascript
Onyx.merge('userWallet', {
    currentBalance: 1000,
    tierName: 'GOLD',
})
```

25. Verify Transfer button is disabled (because there's no selected account)
26. Press Instant (Debit card)
27. Verify two cards appear with correct information
28. Select a card
29. Verify the selected card shows under `Which account?` and Fee shows `$0.25`
30. Press 1-3 Business days (Bank account)
31. Verify two bank accounts appear with correct information
32. Select an account
33. Verify the selected account shows under `Which account?` and Fee shows `$0.00`
34. Press Transfer
35. Verify API is called with correct params
36. Verify error shows `Auth TransferWalletBalance returned an error`

- [x] Verify that no errors appear in the JS console

### Offline tests

NA

#### `ExpensifyCardPage`

 1. On an account with a VIRTUAL and PHYSICAL Expensify card:
 2. Go to `/settings/wallet/card/Expensify`
 3. Verify that the card number is hidden with dots
 4. Press Reveal details
 5. Verify card details is correct; card number is NOT bold; and expiration date is in the format `MM/YYYY`
 
#### `ReportVirtualCardFraudPage`

6. Press Report virtual card fraud >> Deactivate card
7. Verify card is deactivated

#### `ReportCardLostPage`

8. Press Report physical card loss / damage >> Next
9. Verify an error shows require you to choose reason
10. Choose a reason and complete the flow
11. Verify card is reported and closed

#### `AddDebitCardPage`

12. Go to `/settings/wallet/add-debit-card`
13. Fill in invalid values in some fields
14. Verify errors show
15. Leave some fields unfilled >> Next
16. Verify errors show

#### `ActivatePhysicalCardPage`

**Precondition**: Have an unactivated physical card.

17. Go to `/settings/wallet/card/Expensify/activate`
18. Fill in last four digits of card number
19. Press Activate physical card
20. Verify card is activated

#### `TransferBalancePage`, `ChooseTransferAccountPage`

23. Go to `settings/wallet/transfer-balance`
24. Verify balance shows correctly
26. Press Instant (Debit card)
27. Verify two cards appear with correct information
28. Select a card
29. Verify the selected card shows under `Which account?` and Fee shows 1.5% of current balance
30. Press 1-3 Business days (Bank account)
31. Verify two bank accounts appear with correct information
32. Select an account
33. Verify the selected account shows under `Which account?` and Fee shows `$0.00`
34. Press Transfer
35. Verify balance is transfered to the selected account/card

- [x] Verify that no errors appear in the JS console

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If the PR modifies the form input styles:
    - [x] I verified that all the inputs inside a form are aligned with each other.
    - [x] I added `Design` label so the design team can review the changes.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

https://github.com/Expensify/App/assets/113963320/a07bcaec-7095-457f-906a-829ff8e33f89

https://github.com/Expensify/App/assets/113963320/2256cbb9-1b44-43f0-9235-485d26619aae

https://github.com/Expensify/App/assets/113963320/3555e859-b7ca-4624-917e-b19413b04c35

</details>

<details>
<summary>Android: mWeb Chrome</summary>

https://github.com/Expensify/App/assets/113963320/65f27dd9-8485-4e62-abce-6326622442c3

https://github.com/Expensify/App/assets/113963320/1ecaa954-62e2-4802-901b-fca6919324b2

https://github.com/Expensify/App/assets/113963320/269ca72d-564c-4dab-a618-ff9821728723

</details>

<details>
<summary>iOS: Native</summary>

https://github.com/Expensify/App/assets/113963320/4b65be08-9b4d-4534-b794-e0b6a1991495

https://github.com/Expensify/App/assets/113963320/2a616ed8-def1-4569-b939-2627ba158191

https://github.com/Expensify/App/assets/113963320/1bfa970b-a5fc-4fbb-b819-eadf4dc28ff8

https://github.com/Expensify/App/assets/113963320/affd50ed-7a09-4bd3-b67c-c49088adc76e

</details>

<details>
<summary>iOS: mWeb Safari</summary>

https://github.com/Expensify/App/assets/113963320/884110b1-4773-453c-a906-eeeafc5812f6

https://github.com/Expensify/App/assets/113963320/7b5b866f-5645-45f9-8b08-151c80b1e347

https://github.com/Expensify/App/assets/113963320/09c58f1e-9d8b-453a-8d21-bf5a6c27489f

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

#### `ExpensifyCardPage`, `ReportVirtualCardFraudPage`, `ReportCardLostPage`

https://github.com/Expensify/App/assets/113963320/b915e705-ebd7-4603-b285-df8f18817fef

https://github.com/Expensify/App/assets/113963320/4b11184b-83c4-4fe5-a63b-37bf961fb9e1

#### `AddDebitCardPage`

https://github.com/Expensify/App/assets/113963320/84097587-13cd-4150-9c10-31e9af693297

#### `TransferBalancePage`, `ChooseTransferAccountPage`

https://github.com/Expensify/App/assets/113963320/12f049c3-e3be-46cc-9bea-f82764bac506

</details>

<details>
<summary>MacOS: Desktop</summary>

#### `ExpensifyCardPage`, `ReportVirtualCardFraudPage`, `ReportCardLostPage`

https://github.com/Expensify/App/assets/113963320/b915e705-ebd7-4603-b285-df8f18817fef

https://github.com/Expensify/App/assets/113963320/4b11184b-83c4-4fe5-a63b-37bf961fb9e1

#### `AddDebitCardPage`

https://github.com/Expensify/App/assets/113963320/84097587-13cd-4150-9c10-31e9af693297

#### `TransferBalancePage`, `ChooseTransferAccountPage`

https://github.com/Expensify/App/assets/113963320/12f049c3-e3be-46cc-9bea-f82764bac506

</details>

### CODE REVIEW

File: src/pages/settings/Wallet/ReportVirtualCardFraudPage.tsx
Comment (blazejkustra) [inline src/pages/settings/Wallet/ReportVirtualCardFraudPage.tsx:2]: Use isEmptyObject util instead 😄

File: src/pages/settings/Wallet/ChooseTransferAccountPage.tsx
Comment (blazejkustra) [inline src/pages/settings/Wallet/ChooseTransferAccountPage.tsx:24]:
```suggestion
type ChooseTransferAccountPageProps = ChooseTransferAccountPageOnyxProps
```

File: src/pages/settings/Wallet/ChooseTransferAccountPage.tsx
Comment (blazejkustra) [inline src/pages/settings/Wallet/ChooseTransferAccountPage.tsx:21]:
```suggestion
    /** Wallet transfer propTypes */
    walletTransfer: OnyxEntry<WalletTransfer>;
```

File: src/pages/settings/Wallet/ChooseTransferAccountPage.tsx
Comment (blazejkustra) [inline src/pages/settings/Wallet/ChooseTransferAccountPage.tsx:42]:
```suggestion
```

File: src/pages/settings/Wallet/ReportVirtualCardFraudPage.tsx
Comment (blazejkustra) [inline src/pages/settings/Wallet/ReportVirtualCardFraudPage.tsx:26]: Let's add description to these props

File: src/pages/settings/Wallet/ReportVirtualCardFraudPage.tsx
Comment (blazejkustra) [inline src/pages/settings/Wallet/ReportVirtualCardFraudPage.tsx:35]: Let's use StackScreenProps type instead, an example:
```ts
type LogInWithShortLivedAuthTokenPageProps = LogInWithShortLivedAuthTokenPageOnyxProps & StackScreenProps<PublicScreensParamList, typeof SCREENS.TRANSITION_BETWEEN_APPS>;
```

**Review by blazejkustra — CHANGES_REQUESTED** [review]

File: src/pages/settings/Wallet/ActivatePhysicalCardPage.tsx
Comment (filip-solecki) [inline src/pages/settings/Wallet/ActivatePhysicalCardPage.tsx:58]: are you sure it is ok? previously `card.isVirtual` was negated by `!`

File: src/pages/settings/Wallet/ChooseTransferAccountPage.tsx
Comment (filip-solecki) [inline src/pages/settings/Wallet/ChooseTransferAccountPage.tsx:37]: why these decriptions are removed?

File: src/pages/settings/Wallet/ExpensifyCardPage.tsx
Comment (filip-solecki) [inline src/pages/settings/Wallet/ExpensifyCardPage.tsx:63]: I think previous version was not bad - when cardList was not given the function didn't call - currently it calls anyway

File: src/pages/settings/Wallet/ExpensifyCardPage.tsx
Comment (filip-solecki) [inline src/pages/settings/Wallet/ExpensifyCardPage.tsx:89]: why is this assertion needed?

File: src/pages/settings/Wallet/ReportVirtualCardFraudPage.tsx
Comment (filip-solecki) [inline src/pages/settings/Wallet/ReportVirtualCardFraudPage.tsx:45]: same as above

**Comment by situchan** [comment]
TS check failing

File: src/libs/GetPhysicalCardUtils.ts
Comment (situchan) [inline src/libs/GetPhysicalCardUtils.ts:100]: This will cause regression. draftValues can be empty string.

Lets hold this PR for #33641

**Comment by blazejkustra** [comment]
@situchan @tienifr Let's continue work on this one

**Comment by situchan** [comment]
@tienifr ^

File: src/pages/settings/Wallet/ExpensifyCardPage.tsx
Comment (situchan) [inline src/pages/settings/Wallet/ExpensifyCardPage.tsx:80]: This will cause regression if `physicalCard.availableSpend` = 0 and `virtualCard.availableSpend > 0`

File: src/styles/index.ts
Comment (situchan) [inline src/styles/index.ts:4290]: This style was missing before, but I am fine with using default text color and font size.
Just commenting here in case someone reports any style regression pointing to this PR.

**Comment by situchan** [comment]
@tienifr Job3 failing

**Review by situchan — COMMENTED** [review]
Code looks good.

@blazejkustra would you like to do another review and approve?

**Comment by blazejkustra** [comment]
@situchan Yep, go for it!

**Review by blazejkustra — APPROVED** [review]

**Comment by situchan** [comment]
@tienifr I am not able to continue testing on native because of the crash regression [here](https://github.com/Expensify/App/pull/34257/files#r1497361491) caused by your PR

**Comment by situchan** [comment]
> @situchan Do you want me to fix it here?

No, can you create follow-up PR if possible?

**Review by Julesssss — COMMENTED** [review]
> https://github.com/Expensify/App/pull/34716#issuecomment-1956481465

@tienifr please let me know when that PR is ready and I'll review it

**Comment by situchan** [comment]
@tienifr please merge main. Crash is fixed now

**Review by situchan — APPROVED** [review]
Note: we tested using mock data.
Not sure if QA team is able to test this PR properly since they might not have physical card.

**Comment by Julesssss** [comment]
Hey @tienifr, would you mind merging `main` again please, to help prevent non-conflicting issues once merged. Thanks

**Comment by Julesssss** [comment]
> Hey @tienifr, would you mind merging `main` again please, to help prevent non-conflicting issues once merged. Thanks

Actually nevermind, merging main on Friday was probably good enough.

**Review by Julesssss — APPROVED** [review]
I'm struggling to test some of this on dev. Let's get this to staging for QA verification

**Comment by kavimuru** [comment]
@Julesssss @situchan QA team does not have physical card or not activated cards. Is this PR internal?

File: src/pages/settings/Wallet/ReportVirtualCardFraudPage.tsx
Comment (jayeshmangwani) [inline src/pages/settings/Wallet/ReportVirtualCardFraudPage.tsx:48]: Changing the parameter of `getLatestErrorMessage` from `virtualCard` to `virtualCard?.errors ?? {}` causes an issue where the server error does not display. This happens because the updated parameter always returns an empty string more context on this Issue https://github.com/Expensify/App/issues/49010#issuecomment-2345118755.

**Pull request merged**
